docs/2020-05-22-Merging-Rules.md

```diff
@@ -0,0 +1,221 @@
+# Background
+
+K rules, are of the form
+
+    rule l => r requires c
+
+where, in general, l is a constructor-based term
+(modulo axioms: may contain Ac constructors, overloadable constructors)
+
+When translated into matching logic, they are of the form
+
+    axiom l /\ c -> EX r
+
+Where EX is the next operator, meaning that there is a possible transition
+from l (satisfying c) to r.
+
+# Approach
+
+We will only discuss here combining two rules, but it should be clear
+that this easily extends to any finite number of rules.
+
+## FOL interlude
+
+The following is provable using FOL reasoning:
+
+```coq
+Theorem combine
+  (p1 q1 p2 q2 : var -> Prop)
+  (i j : nat)
+  (Hnij : i <> j)
+  : (forall x, p1  x -> q1 x) /\ (forall y, p2 y-> q2 y) 
+    <-> forall (c : nat) x y,
+        (c = i /\ p1 x) \/ (c = j /\ p2 y)
+        -> (c = i /\ q1 x) \/ (c = j /\ q2 y).
+```
+
+## Back to ML
+
+Given two rules
+
+    axiom forall x1 . l1 -> EX r1
+    axiom forall x2 . l2 -> EX r2
+
+From the theorem above, we can combine the two axiom into an equivalent one:
+
+    axiom forall c x1 x2 .
+        (c = 1 /\ l1) \/ (c = 2 /\ l2) ->
+        (c = 1 /\ EX r1) \/ (c = 2 /\ EX r2)
+
+
+Given that `c = 1` and `c = 2` are predicates, and that EX is a standard symbol,
+we can transform the RHS of the rule into
+
+    EX ((c = 1 /\ r1) \/ (c = 2 /\ r2))
+
+## Side conditions
+
+If `l1 = tl1 /\ pl1` and `l2 = tl2 /\ pl2` then we can transform the lhs above
+into `((c = 1 /\ tl1) \/ (c = 2 /\ tl2)) /\ ((c = 1 /\ tp1) \/ (c = 2 /\ tp2))`
+
+Then, using the distributivity of `/\` over `\/` we can factor out common 
+predicates from the side condition.
+
+## Pushing choice predicates into terms
+
+
+In fact, the property we stated above for `EX` holds for any symbol in the
+signature, so we can easily prove that
+
+    (c = 1 /\ f(t11, .., t1n)) \/ c = 2 /\ f(t21, ..., t2n))
+
+is equivalent to
+
+    f((c = 1 /\ t11) \/ (c = 2 /\ t21), ..., (c = 1 /\ t1n) \/ (c = 2 /\ t2n))
+
+Note 1: in the case we have the same term / variable / constant in both rules,
+we can transform `(c = 1 /\ t) \/ (c = 2 /\ t)` into `t /\ (c = 1 \/ c = 2)`
+
+Note 2: When reaching an builtin construct we might choose to not push choices
+further on.  If we do, then builtin unification would have to handle choices as
+well.
+
+### Indexed normal form
+
+Applying the above we can compress the two rules to something like:
+
+    axiom forall x1 x2 c .
+      G[(c = 1 /\ t11) \/ (c = 2 /\ t21), ..., (c = 1 /\ t1m) \/ (c = 2 /\ t2m)]
+      =>
+      G'[(c = 1 /\ t11') \/ (c = 2 /\ t21'), ..., (c = 1 /\ t1n') \/ (c = 2 /\ t2n')]
+
+Where `G` and `G'` are multi-hole contexts such that:
+
+- `G[t11, ..., t1m] = l1` and `G[t21, ..., t2m] = l2`,
+- `symbol(t1i) != symbol(t2i)` for each `1 <= i <= m`
+- `G'[t11', ..., t1n'] = r1` and `G'[t21', ..., t2n'] = r2`,  
+- `symbol(t1j') != symbol(t2j')` for each `1 <= j <= n`
+
+
+Note 1: the restriction on top symbols are optional; we might not want to
```

Comment:
Actually I'm not so sure about the top symbol constraints. Can they always be satisfied by some G and G'?